myia/opt.py

```diff
@@ -0,0 +1,205 @@
+"""Graph optimization routines."""
+
+
+from myia.unify import Var, Unification, expandlist
+from myia.graph_utils import toposort
+from myia.anf_ir import ANFNode, Graph, Apply, Constant, Special
+from myia.anf_ir_utils import \
+    succ_incoming, freevars_boundary, replace, \
+    is_apply, is_constant, is_parameter
+
+
+class VarNode(Special):
+    """Graph node that represents a variable."""
+
+    def __var__(self):
+        return self.special
+
+
+def sexp_to_node(sexp, graph, multigraph=False):
+    """Convert an s-expression (tuple) to a subgraph.
+
+    Args:
+        sexp: A nested tuple that represents an expression.
+        graph: The graph in which to place the created nodes.
+        multigraph: If multigraph is True and graph is a Var, then
+            every child node will have a fresh Var as its Graph.
+            In short, use multigraph=True to get a subgraph where
+            each node can be in a different graph. Otherwise, all
+            nodes are required to belong to the same graph.
+
+    Returns:
+        An ANFNode equivalent to the given s-expression.
+
+    """
+    if isinstance(sexp, tuple):
+        if multigraph and isinstance(graph, Var):
+            return Apply([sexp_to_node(x, Var('G'), True)  # type: ignore
+                          for x in sexp], graph)  # type: ignore
+        else:
+            return Apply([sexp_to_node(x, graph, multigraph)
+                          for x in sexp], graph)
+    elif isinstance(sexp, Var):
+        return VarNode(sexp, graph)
+    elif isinstance(sexp, ANFNode):
+        return sexp
+    else:
+        return Constant(sexp)
+
+
+def sexp_to_graph(sexp):
+    """Convert an s-expression to a Graph.
+
+    This converts the s-expression to a subgraph of ANFNodes and then sets that
+    subgraph as the output of a new Graph.
+    """
+    g = Graph()
+    g.output = sexp_to_node(sexp, g)
+    return g
+
+
+class GraphUnification(Unification):
+    """Unification subclass that can unify subgraphs."""
+
+    def visit(self, fn, node):
+        """Visit a node and apply fn to all children."""
+        if is_apply(node):
+            new_inputs = expandlist(map(fn, node.inputs))
+            g = fn(node.graph)
+            return Apply(new_inputs, g)  # type: ignore
+        elif is_parameter(node):
+            g = fn(node.graph)
+            if not isinstance(g, Graph) or g is not node.graph:
+                # Note: this condition will be triggered if e.g. there is a
+                # Parameter in a pattern to reify. It's not clear what that's
+                # supposed to mean unless the Parameter already exists in a
+                # concrete graph, so we raise an Exception just in case.
+                raise Exception('Unification cannot create new Parameters.') \
+                    # pragma: no cover
```

Comment:
Why are you doing this check?

In any case, if you need to keep it, I would make that an assert.

Reply:
Because when I was testing things, I carelessly made a pattern that had a parameter, and it took me a while to realize why matching didn't work. I think I prefer to keep an Exception (maybe TypeError would be more appropriate), in case this (constructing patterns) gets exposed to more developers in the future.